ghcide/src/Development/IDE/Core/ProgressReporting.hs

```diff
@@ -102,6 +111,12 @@ recordProgress InProgressState{..} file shift = do
         return (prev, new)
     alter x = let x' = maybe (shift 0) shift x in Just x'
 
+-- | Runs the action until it ends or until the given MVar is put.
+--   Rethrows any exceptions.
+untilMVar :: MonadUnliftIO m => MVar () -> m () -> m ()
+untilMVar mvar io = void $
+    waitAnyCancel =<< traverse async [ io , readMVar mvar ]
```

Comment:
I'm not using async stuff very often, so take this suggestion with a grain of salt.
But I think you can write this using the more higher-level function [race_](https://hackage.haskell.org/package/async-2.2.4/docs/Control-Concurrent-Async.html#v:race_) (resp. its unliftio counterpart)

Reply:
That is simpler, but [race_](https://hackage.haskell.org/package/async-2.2.4/docs/Control-Concurrent-Async.html#v:race_) would hide the exception?🤔

Reply:
What exception? And do you or do you not want that exception to be hidden?

Reply:
I do not want exception to be hidden if any.

Reply:
In that case race_ should be just fine, because it [rethrows](https://hackage.haskell.org/package/async-2.2.4/docs/src/Control.Concurrent.Async.html#race) the exception if any of the 2 actions throws any.